Added CW-SSIM for computing SSIM in complex wavelet domain

chaosmail opened the pull request:
Hi guys,

thanks for your nice library! I added SSIM computation in complex wavelet domain (IMAGE SIMILARITY IN COMPLEX WAVELET DOMAIN from Zhou Wang and Eero P. Simoncelli), which is invariant to translation and rotation. I added an IPhython notebook with the examples and benchmarks.

Best,
Christoph

jterrace commented:
Could you fix the pylint errors? See the README for how to run it.

jterrace commented on .gitignore:
```diff
@@ -0,0 +1,3 @@
+*.pyc
```
These should be in your global .gitignore, rather than the project one.
See: https://help.github.com/articles/ignoring-files/#create-a-global-gitignore

chaosmail replied:
alright, I removed the .gitignore again

jterrace commented on ssim/ssimlib.py:
```diff
@@ -34,7 +37,6 @@ def __init__(self, img, gaussian_kernel_1d, size=None):
           gaussian_kernel_1d: Gaussian kernel from get_gaussian_kernel.
```
please update docstring

chaosmail replied:
ok :)

jterrace commented on ssim/ssimlib.py:
```diff
@@ -45,25 +47,30 @@ def __init__(self, img, gaussian_kernel_1d, size=None):
 
         Args:
           img1: Reference image to compare other images to.
           gaussian_kernel_1d: Gaussian kernel from get_gaussian_kernel.
```
update docstring here too

chaosmail replied:
also :)

jterrace commented on ssim/ssimlib.py:
```diff
@@ -107,6 +114,61 @@ def ssim_value(self, target):
+          Computed CW-SSIM float value.
+        """
+        if isinstance(target, compat.basestring):
+            target = compat.Image.open(target)
```
why not use SSIMImage here?

chaosmail replied:
good point, I restructure this a little

jterrace commented:
Could you also add a command-line flag to use this new algorithm and add a test that uses it in `.travis.yml`?

chaosmail commented:
Alright, I tried to address all your remarks in the last commit.

chaosmail commented:
hmm what's going on in the Travis build, there are strange errors. I also added Python 3.4 to the .travis.yml because I am using it on my machine and it works nice; however, there is also some issue in Travis.

chaosmail commented:
Alright, I could resolve all the errors; I rebased the PR to a nice commit. Btw. I added conda to Travis, to speed up the tests; also I removed python 3.2 (it should not be used) and added python 3.3 and 3.4 to Travis. I added tests, some description in the README, some command line arguments and the Reference to the paper. Let me know if you need further adjustments ;)

chaosmail commented:
I almost forgot; there is also the IPython file in the root folder (comparing the 2 methods) and the PDF paper in the readings folder. Hope that's ok - if not then I will remove them.
Best,
Christoph

jterrace commented:
Yeah, 3.2 was broken at head too. You actually managed to fix the errors? Sweet!

jterrace commented on .travis.yml:
```diff
@@ -1,23 +1,28 @@
 
 before_install:
- - sudo apt-get install -qq python-numpy python-scipy python3-numpy python3-scipy
+  - wget http://repo.continuum.io/miniconda/Miniconda-latest-Linux-x86_64.sh -O miniconda.sh
```
can you change this to https?

jterrace commented on .travis.yml:
```diff
@@ -1,23 +1,28 @@
-  - python setup.py install
+  - conda create --yes -q -n pyenv python=$TRAVIS_PYTHON_VERSION pillow numpy scipy nose pylint pip
+  - source activate pyenv
+  # install 
```
kind of a silly comment :P

jterrace commented on ssim/ssimlib.py:
```diff
@@ -118,22 +202,42 @@ def main():
         prog='pyssim', formatter_class=argparse.RawTextHelpFormatter,
         description=description)
+    parser.add_argument('--cw', help='compute the complex wavelet SSIM',
+                        action="store_true")
```
single quotes to match rest of the file

jterrace commented:
Just two minor comments and it's ready to merge. Thanks for the contribution!

chaosmail commented:
alright, here we go.

jterrace commented:
Nice contribution. Thank you!

jterrace commented:
Once thing I forgot - will you send me a followup CL to add yourself to the AUTHORS file?
